### PR TITLE
client: expose SoftLayerClient so we can make us of it

### PR DESCRIPTION
**Pull request by fatih**

### CODE REVIEW

**Comment by fatih** [comment]
Right now I can't store the client for example to a field of a certain struct. This makes developing hard, it requires us to create the client everytime we want to use it.

**Comment by maximilien** [comment]
@fatih, Tell me more why you want to do this. I purposely made the struct private since it keeps sensitive information (`username` and `apiKey`) which I do not want to expose. 

Creating and deleting the client object is anyhow cheap. Further, you should be using the interface type if you wanted to keep the client around, e.g.,

```
import (
  slclient github.com/maximilien/softlayer-go/client
  softlayer github.com/maximilien/softlayer-go/softlayer
)

var client softlayer.Client

...
client = slclient.NewSoftLayerClient(username, apiKey)
...

```

The client var can be kept around as needed. I am closing this, but if you have a strong argument then comment and reopen it.

Take care,

Max

**Comment by fatih** [comment]
The client interface is a very large interface, and I think it's the wrong way to use it and not Go idiomatic. But if you believe it should be that way, I can't say much. 

I need to store so I don't create every time a new client. This is imho important and every single other provider API let me do this. And I don't believe it's cheap because every time `NewSoftLayerClient` will do the followings:

* Call `os.Getwd()`, which the error is neglected and imho should be handled. Right now we don't handle it and it's bad
* Creates an `*http.Client` which is overwritten every time we create another client instance. We should expose that field so we can pass custom http.Fields, just like AWS and others do it.
* We allocate a map (`softlayerServices`) and initialize it with all services.

Having said that, we'll going to create a client for  approx 10req/sec in our production environment. The listed problems above means this is costly (allocation), not safe (os.Getwd() is not checked) and also not extensible (no way to pass custom http client, for example with a custom timeout)

The ideal way would be initialize a single service once and just use it for every single interaction.

Also making it private doesn't solve the security issue, because it's the responsibility of the package user. This is not exposed in anyway to a third party service. This package will be used by developers to, it's not something that is consumed by third party users in any way. Please reconsider my comment and also have a look at other third party packages of other providers. This code right now is not idiomatic Go and it makes developing hard, not easy.

**Comment by fatih** [comment]
@maximilien another issue is that I also need a custom `*http.Client` which is right now not possible. I know I said it above, but I just wanted to say it again. Can you please reconsider your decisions or rethink for a better solution. I would appreciate it. Thanks.

**Comment by maximilien** [comment]
OK, I can buy the need for custom `HttpClient` especially if you are looking at such a high request env.

Can you update this PR with an injected `HttpClient` or some similar solution that would end up being helpful for you? Don't forget to update tests.

Thanks,

Max

**Comment by fatih** [comment]
Thanks @maximilien for reconsidering. I've done the followings:

* Exposed `httpClient` and renamed it to `HTTPClient`. I've also defined a `DefaultHTTPClient` which we use. In the future we might modify this based on our requirements if needed.
* We also panic for `os.Getwd()` to prevent any kind of malformed API calls when it fails. When `os.Getwd()` fails it's usually a system problem which needs to be fixed before the customer of this API can move forward. I didn't change the signature of `*SoftlayerClient` for backwards compatibility.

File: client/softlayer_client.go [hunk 1]
```diff
@@ -16,39 +16,42 @@ import (
 	softlayer "github.com/maximilien/softlayer-go/softlayer"
 )
 
+var DefaultHTTPClient = http.DefaultClient
```
Comment (maximilien) [inline client/softlayer_client.go]: Why create an additional `NewSoftLayerClient` where the `HttpClient` is injected and this one by default uses the `http.DefaultClient`?

Since this one `NewSoftLayerClient(username, apiKey string) *SoftLayerClient` would still exist, current client satisfied with default would be OK. That way it's the responsibility of the caller to either create the HTTP client and pass it or use the default?

If you are OK with that please add some tests for both `NewSoftLayerClient` functions and rebase and I will merge.

Sorry about back and forth, want to make sure we have something we all like.

Thanks much. Best.

**Comment by fatih** [comment]
@maximilien I've rebased and change the way HTTPClient is passed. I also added a test. Please let me know if this is NOT what you're expecting. I'm happy to fix/change it further. Thanks.

File: client/softlayer_client.go [hunk 2]
```diff
@@ -22,33 +22,34 @@ const (
...
-type softLayerClient struct {
+type SoftLayerClient struct {
 	username string
 	apiKey   string
 
 	templatePath string
 
-	httpClient *http.Client
+	// HTTPClient is used for HTTP requests. By default DefaultHTTPClient is
+	// being used.
```
Comment (maximilien) [inline client/softlayer_client.go]: No need for comments. Code is clear. Comment get out of date :)

File: client/softlayer_client.go [hunk 3]
```diff
@@ -22,33 +22,34 @@ const (
...
 	nonVerbose bool
 }
 
-func NewSoftLayerClient(username, apiKey string) *softLayerClient {
-	pwd, _ := os.Getwd()
-	slc := &softLayerClient{
+func NewSoftLayerClient(username, apiKey string) *SoftLayerClient {
+	pwd, err := os.Getwd()
+	if err != nil {
+		panic(err) // this should be handled by the user
```
Comment (maximilien) [inline client/softlayer_client.go]: Why not return it... the code that uses `NewSoftLayerClient` can then decide. I realize this breaks interface but it's a change I am willing to make. Thoughts?

File: client/softlayer_client_test.go [hunk 4]
```diff
@@ -35,6 +35,16 @@ var _ = Describe("SoftLayerClient", func() {
 		})
 	})
 
+	Context("#NewSoftLayerClient_HTTPClient", func() {
```
Comment (maximilien) [inline client/softlayer_client_test.go]: Yes. This is good. However, I'd expect another `It()` that passes your own `HTTPClient` and verifies that it is indeed used. If you add that should be good enough coverage. Thanks.

File: client/softlayer_client.go (hunk 3 above)
Comment (fatih) [inline client/softlayer_client.go]: I would do the same and I agree with you, but I would do this with another PR. Because this requires us to change and rethink all the tests and the current usages of this constructor. Also I think it's better in long term to improve the state of this package.

File: client/softlayer_client.go [hunk 5]
```diff
@@ -22,33 +22,34 @@ const (
...
+	}
+
+	slc := &SoftLayerClient{
 		username: username,
 		apiKey:   apiKey,
 
 		templatePath: filepath.Join(pwd, TEMPLATE_ROOT_PATH),
```
Comment (maximilien) [inline client/softlayer_client.go]: I am even wondering if we really need that `templatePath` anymore... Could we get rid of it?

File: client/softlayer_client.go (hunk 3 above)
Comment (maximilien) [inline client/softlayer_client.go]: Actually I think we can remove the call for `pwd, err := os.Getwd()` since it's used to set `templatePath` which is only used in `GenerateRequestBody` and I think that's not used much anywhere.

I have to double check. So leave the `panic(err)` and I will open an issue to investigate removing `templatePath` and `GenerateRequestBody` from interface. Do not use them since they will likely be removed.

Thanks.

File: client/softlayer_client_test.go (hunk 4 above)
Comment (fatih) [inline client/softlayer_client_test.go]: Done. I've added the testing for a custom HTTP client. I've rebased and force pushed, please have check: 

https://github.com/fatih/softlayer-go/commit/d395a37d9d93532cc41c93e864be3101dd5b29fe#diff-7ad39016e99d3c1c7dab0d9b0245f116R50

**Comment by fatih** [comment]
@maximilien I've fixed all concerns and comments. Please let me know if something is missing. Thanks

**Comment by maximilien** [comment]
Hi, @fatih, sorry about the delay. I was traveling to Shanghai and finally back into full work mode now. I will review soon. I want to apologize for not mentioning this prior.

**Comment by fatih** [comment]
@maximilien no worries, happy you're back :) Please let me know if something is wrong. Also we at Koding are using it and planning to add non existing features to it in couple of weeks. Thanks.

**Comment by maximilien** [comment]
@fatih wonderful to know. 

Please feel free to send separate PRs for the new features. 

I too will likely add more support for terribly named `Virtual_Guest_Block_Device_Template_Group_Service` service in the near future.

File: client/softlayer_client.go (hunk 5 above)
Comment (maximilien) [inline client/softlayer_client.go]: I opened issue #74 to remove the `templatePath` and associated potential dead code?

**Comment by maximilien** [comment]
OK, I like it. Merging. Cheers :beers: 

/cc @jianqiu please let me know if you see anything wrong with this change in our pipeline. In future we can use the `develop` branch to monitor changes instead of asking you.
